RestClient.php

```diff
@@ -1,64 +1,135 @@
+        Response::HTTP_GONE,
+        
+    ];
+
+    /**
+     * RestClient constructor
+     */
+    public function __construct() {
+        $this->restClient = (new RestClientFactory())->createOne([]);
+    }
+
+    /**
+     * sends the request
+     *
+     * @param  Request $request
+     * @return Response
+     * @throws RequestFailedException
+     *
+     * @SuppressWarnings("PHPMD.StaticAccess")
+     */
+    public function send(Request $request) {
+        $method     = strtolower($request->getMethod());
+        $response   = $method === HttpMethods::GET || $method === HttpMethods::DELETE ? $this->restClient->$method($request->getUrlAndQuery(), $request->getCurlOptions()) : $this->restClient->$method($request->getUrlAndQuery(), $request->getPayload(), $request->getCurlOptions());
+
+        try {
+            return $this->isSuccessfulRequest($response->getStatusCode()) ? $response : Exceptions::RequestFailedException($request, $response->getStatusCode(), $response->getContent());
+        } catch (DBALException\DriverException $e) {
+            throw $this->handleFailedResponse($response, $e);
+        }
+    }
+    
+    /**
+     * Handle a failed response by creating a specific exception for the given
+     * response.
+     * 
+     * @param Response $response
+     * @param DriverExceptionInterface $exception
+     * @return DriverException
+     */
+    private function handleFailedResponse(Response $response, DriverExceptionInterface $exception) {
+        switch ($response->getStatusCode()) {
+            case Response::HTTP_CONFLICT:
+                return new DBALException\UniqueConstraintViolationException($response->getContent(), $exception);
+
+            case Response::HTTP_BAD_REQUEST:
+                return new DBALException\SyntaxErrorException($response->getContent(), $exception);
+
+            case Response::HTTP_METHOD_NOT_ALLOWED:
+            case Response::HTTP_NOT_ACCEPTABLE:
+            case Response::HTTP_REQUEST_TIMEOUT:
+            case Response::HTTP_LENGTH_REQUIRED:
+            case Response::HTTP_INTERNAL_SERVER_ERROR:
+                return new DBALException\ServerException($response->getContent(), $exception);
+            
+            case Response::HTTP_CONFLICT:
+                return new DBALException\ConstraintViolationException($response->getContent(), $exception);
+
+            case Response::HTTP_UNAUTHORIZED:
+            case Response::HTTP_FORBIDDEN:
+            case Response::HTTP_PROXY_AUTHENTICATION_REQUIRED:
+            case Response::HTTP_BAD_GATEWAY:
+            case Response::HTTP_SERVICE_UNAVAILABLE:
+            case Response::HTTP_GATEWAY_TIMEOUT:
+                return new DBALException\ConnectionException($response->getContent(), $exception);
+
+            default:
+                return new DBALException\DriverException($response->getContent(), $exception);
+
+        }
+    }
+    
+    private function isSuccessfulRequest($statusCode)

```

Comment:
I don't think that this is the best idea. Imagine a GET request with response code 204 no content. I don't think that's a valid response code.

You exception handling is awesome, handling success codes needs much more detail and it has to stay configurable I think. Opinion on that?